pkg/uri/parser.go

```diff
@@ -51,118 +46,87 @@ func (u *URI) IsFullyQualified() bool {
 	return len(u.Repository) > 0 && len(u.Ref) > 0 && u.Path != nil
 }
 
+func (u *URI) GetPath() string {
+	if u.Path == nil {
+		return ""
+	}
+	return *u.Path
+}
+
 func (u *URI) String() string {
 	var buf strings.Builder
 	buf.WriteString(u.Protocol)
 	buf.WriteString(ProtocolSeparator)
 	buf.WriteString(u.Repository)
-
 	if len(u.Ref) == 0 {
 		return buf.String()
 	}
-	buf.WriteRune(RefSeparator)
+	buf.WriteString(PathSeparator)
 	buf.WriteString(u.Ref)
-
 	if u.Path == nil {
 		return buf.String()
 	}
-	buf.WriteRune(PathSeparator)
+	buf.WriteString(PathSeparator)
 	buf.WriteString(*u.Path)
-
 	return buf.String()
 }
 
-func Parse(str string) (*URI, error) {
-	// start with protocol
-	protoParts := strings.Split(str, ProtocolSeparator)
-	const uriParts = 2
-	if len(protoParts) != uriParts {
+// ParseWithBaseURI parse URI uses base URI as prefix when set and input doesn't start with lakeFS protocol
+func ParseWithBaseURI(s string, baseURI string) (*URI, error) {
+	if len(baseURI) > 0 && !strings.HasPrefix(s, LakeFSProtocol+ProtocolSeparator) {
+		s = baseURI + s
+	}
+	u, err := Parse(s)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", err, s)
+	}
+	return u, nil
+}
+
+func Parse(s string) (*URI, error) {
+	u, err := url.Parse(s)
+	if err != nil || u.Scheme != LakeFSProtocol {
 		return nil, ErrMalformedURI
 	}
-	if !strings.EqualFold(protoParts[0], LakeFSProtocol) {
+	repository := u.Hostname()
+	if len(repository) == 0 {
 		return nil, ErrMalformedURI
 	}
-
-	var uri URI
-	uri.Protocol = protoParts[0]
-	var path string
-
-	var state = stateInRepo
-	var buf strings.Builder
-	for _, ch := range protoParts[1] {
-		switch {
-		case ch == RefSeparator && state == stateInRepo:
-			uri.Repository = buf.String()
-			state = stateInRef
-			buf.Reset()
-		case ch == PathSeparator && state == stateInRef:
-			uri.Ref = buf.String()
-			state = stateInPath
-			uri.Path = &path
-			buf.Reset()
-		default:
-			buf.WriteRune(ch)
+	var ref string
+	var path *string
+	if len(u.Path) > 0 {
+		if !strings.HasPrefix(u.Path, "/") {
+			return nil, ErrMalformedURI
 		}
```

Comment:
Is this even possible?
Skip this check if it isn't.

Reply:
`lakefs://a`